src/OpenLoco/src/World/TownManager.cpp

```diff
@@ -20,6 +20,112 @@ namespace OpenLoco::TownManager
 {
     static loco_global<Town*, 0x01135C38> _dword_1135C38;
 
+    // 0x0049B45F
+    static uint32_t calcCargoInfluenceFlags()
+    {
+        registers regs;
+        call(0x0049B45F, regs);
+        return regs.eax;
+    }
+
+    // 0x004978B7
+    static bool generateTownName(Town* town)
+    {
+        registers regs;
+        regs.esi = X86Pointer(town);
+        return !(call(0x004978B7, regs) & X86_FLAG_CARRY);
+    }
+
+    static auto& rawTowns() { return getGameState().towns; }
+
+    // 0x00496FE7
+    Town* initialiseTown(World::Pos2 pos)
+    {
+        Town* town = nullptr;
+        for (auto& candidateTown : rawTowns())
+        {
+            if (candidateTown.empty())
+            {
+                town = &candidateTown;
+                break;
+            }
+        }
+
+        // No space for a new town?
+        if (town == nullptr)
+        {
+            return nullptr;
+        }
+
+        // Initialise the new town
+        town->x = pos.x;
+        town->y = pos.y;
+        town->flags = TownFlags::none;
+        town->population = 0;
+        town->populationCapacity = 0;
+        town->numBuildings = 0;
+        town->size = TownSize::hamlet;
+        town->historySize = 1;
+        town->history[0] = 0;
+        town->historyMinPopulation = 0;
+
+        std::fill_n(&town->var_150[0], std::size(town->var_150), 0);
+
+        town->var_19C[0][0] = 0;
+        town->var_19C[0][1] = 0;
+        town->var_19C[1][0] = 0;
+        town->var_19C[1][1] = 0;
+        town->numStations = 0;
+        town->var_1A8 = 0;
+
+        town->prng = getGameState().rng;
+
+        std::fill_n(&town->companyRatings[0], std::size(town->companyRatings), 500);
+
+        town->companiesWithRating = 0;
+
+        std::fill_n(&town->monthlyCargoDelivered[0], std::size(town->monthlyCargoDelivered), 0);
+
+        town->cargoInfluenceFlags = calcCargoInfluenceFlags();
+        town->buildSpeed = 1;
+
+        // Figure out a name for this town?
+        if (!generateTownName(town))
+        {
+            town->name = StringIds::null;
+            return nullptr;
+        }
+
+        // Figure out if we need to reset building influence
+        for (auto& otherTown : rawTowns())
+        {
+            if (otherTown.empty())
+            {
+                continue;
+            }
+
+            if (otherTown.numBuildings > 0)
```

Comment:
i think you have these checks back to front. Its meant to be **don't** `resetBuildingsInfluence` if numBuildings == 0 and population == 0 and popCapacity == 0. 

The way i see this working is that ideally you should always `resetBuildingsInfluence` when a new town is created as that will assign the influence to the correct town. BUT when adding towns in scenario editor you are adding multiple towns one after each other so its pointless to `resetBuildingsInfluence` on each add and you should just do it at the end after all towns are added. So i think that is what this code is for its an optimisation for the scenario editor.

Reply:
Ahh, that makes much more sense. I've changed the logic to match, thanks!